[bugfix] check whether MoE Layers are ignored by quantization

HDCharles opened the pull request:
Applying quantization while ignoring some MoE layers but not others would cause model load errors due to vllm assuming all the layers were quantized since there was no check of the ignore list unlike the one for [linear](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py#L138)

Changes
1) previously the code had to [duplicate the get_scheme functionality for FusedMoE](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py#L113-L136) due to layer name changes and the layer no longer being linear. This duplicated code had issues like not taking into account the ignore list. I got rid of the duplication by making it so FusedMoE can use get_scheme if we modify the CTconfig slightly.
2) refactor the get_moe_method to use quant_scheme rather than quant_config, 
2a)^ requires changing a lot of non-static CompressedTensorConfig methods like _is_wNa16_group_channel to be converted to static/class methods since we won't have access to teh quant_config to call them. (Is this fine? TODO)
3) we check for qschemes of all 3 projections but for only expert 0, this aligns with the old logic which used a regex and only returned the first match but could be corrected. It may be slow though.

depends on [28770](https://github.com/vllm-project/vllm/pull/28870) and [28771](https://github.com/vllm-project/vllm/pull/28871)

@kylesayrs @dsikka @brian-dellabetta

HDCharles committed:
Actually Ignore non-quantized MoE Layers

Applying quantization to some MoE layers but not others would cause
model load errors due to vllm assuming all the layers were quantized
since there was no check of the ignore list unlike the one for linear
(https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py#L138)

TODO
add test
update TODO

Signed-off-by: HDCharles <charlesdavidhernandez@gmail.com>

HDCharles commented on vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:
```diff
@@ -106,41 +106,13 @@ class CompressedTensorsMoEMethod(FusedMoEMethodBase):
     def get_moe_method(
         quant_config: "CompressedTensorsConfig",  # type: ignore # noqa E501
```
we can get rid of quant_config entirely here since we have quant_scheme but it would require us converting 5 or so CompressedTensorsConfig.is_X methods to class/static methods at which point i'd probably just change them all. Want to make sure i'm on the right track before i start doing all that.

kylesayrs commented on vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:
```diff
@@ -158,9 +158,36 @@ def get_quant_method(
+                    "All MoE projections need to have same quantization scheme"
+                )
+
+            quant_scheme = all_quant_schemes.pop()
```
I think this logic should go in `CompressedTensorsMoEMethod.get_moe_method`

HDCharles replied:
shouldn't it match the way we're doing it for linear?

HDCharles committed:
CTConfig methods to static/class methods

Summary:

These methods which accept QuantizationArgs already have the necessary
information passed in and don't require the whole
CTConfig. Converting these methods to static and class methods allows
us to avoid carrying the whole CTConfig through all of the MoEMethod
calculation which will be fixed in another PR

Signed-off-by: HDCharles <charlesdavidhernandez@gmail.com>

HDCharles committed:
[refactor] CTMoEMethod Calls

Summary:

The various MoEMethods don't need the entire CTConfig, they
only use the QuantizationArgs from the layer being quantized. Changing
this allows us to continue to remove usage of the entire CTConfig.

Depends on https://github.com/vllm-project/vllm/pull/28870

Signed-off-by: HDCharles <charlesdavidhernandez@gmail.com>
